geoportal/c2cgeoportal_geoportal/lib/caching.py

```diff
@@ -258,16 +262,19 @@ def _set_common_headers(
 
     response.headers.update(service_headers_settings.get("headers", {}))
 
+    if cache in (Cache.PRIVATE, Cache.PRIVATE_NO, Cache.PUBLIC_CAN_PRIVATE, Cache.PUBLIC_CAN_PRIVATE_NO):
+        response.vary = (response.vary or ()) + ("Cookie",)
+
     maxage = (
         service_headers_settings.get("cache_control_max_age", 3600)
-        if cache in (Cache.PUBLIC, Cache.PRIVATE)
+        if cache in (Cache.PUBLIC, Cache.PUBLIC_CAN_PRIVATE, Cache.PRIVATE)
         else service_headers_settings.get("cache_control_max_age_nocache", 10)
     )
     response.cache_control.max_age = maxage
     if maxage == 0:
         response.cache_control.no_cache = True
         response.cache_control.no_store = True
-    elif cache in (Cache.PUBLIC, Cache.PUBLIC_NO):
+    elif cache in (Cache.PUBLIC, Cache.PUBLIC_CAN_PRIVATE, Cache.PUBLIC_NO, Cache.PUBLIC_CAN_PRIVATE_NO):
```

Comment:
In case PUBLIC_CAN_PRIVATE, if user is authenticate we should return private, so it is the same as PRIVATE.